### ISSUE TITLE
REPL on WSL takes long time to start

### ISSUE THREAD

**Issue by sailikk** [greghendershott/racket-mode#502]
Hello! I am on WSL Ubuntu-18.04 LTS build using bash. When I first tried to run the REPL, it wouldn't start at all! So I did some digging around here and `raco pkg install drracket` seemed to fix that issue. 

However, one issue that is still bugging me is the 15-20 second startup time for the REPL. After doing `C-c C-c` I must wait at least 15 seconds for the REPL split window to show up. Compared to the "run in shell using `racket`" option, which is almost instantaneous, I was wondering if this startup time is even normal? I have never used Linux on anything but Windows so I have no experience on what it is with other systems, but this at least seems unusual.

`racket-bug-report` output:
<details>
```
((alist-get 'racket-mode package-alist)
 #s(package-desc racket-mode
                 (20201022 1304)
                 "Racket editing, REPL, and more"
                 ((emacs
                   (25 1))
                  (faceup
                   (0 0 2))
                  (pos-tip
                   (20191127 1028)))
                 nil nil "/home/sailik/.emacs.d/elpa/racket-mode-20201022.1304"
                 ((:url . "https://www.racket-mode.com/")
                  (:maintainer "Greg Hendershott")
                  (:authors
                   ("Greg Hendershott"))
                  (:commit . "41abea866709c5eb3b48f1876aa5dedc75080d6b"))
                 nil))
((emacs-version "27.1")
 (system-type gnu/linux)
 (x-gtk-use-system-tooltips t)
 (major-mode racket-mode)
 (racket--el-source-dir "/home/sailik/.emacs.d/elpa/racket-mode-20201022.1304/")
 (racket--rkt-source-dir "/home/sailik/.emacs.d/elpa/racket-mode-20201022.1304/racket/")
 (racket-program "/usr/local/bin/racket")
 (racket-command-timeout 10)
 (racket-xp-after-change-refresh-delay 1)
 (racket-xp-highlight-unused-regexp "^[^_]")
 (racket-repl-buffer-name-function nil)
 (racket-memory-limit 2048)
 (racket-error-context medium)
 (racket-history-filter-regexp "\\`\\s *\\'")
 (racket-images-inline t)
 (racket-images-keep-last 100)
 (racket-images-system-viewer "display")
 (racket-images-system-viewer "display")
 (racket-use-repl-submit-predicate nil)
 (racket-pretty-print t)
 (racket-indent-curly-as-sequence t)
 (racket-indent-sequence-depth 0)
 (racket-pretty-lambda nil)
 (racket-smart-open-bracket-enable nil)
 (racket-module-forms "\\s(\\(?:module[*+]?\\|library\\)")
 (racket-logger-config
  ((cm-accomplice . warning)
   (GC . info)
   (module-prefetch . warning)
   (optimizer . info)
   (racket/contract . error)
   (sequence-specialization . info)
   (* . fatal)))
 (racket-show-functions
  (racket-show-pseudo-tooltip)))
(enabled-minor-modes
 (auto-compile-on-load-mode)
 (auto-composition-mode)
 (auto-compression-mode)
 (auto-encryption-mode)
 (auto-fill-mode)
 (auto-save-mode)
 (blink-cursor-mode)
 (electric-indent-mode)
 (evil-local-mode)
 (evil-mode)
 (file-name-shadow-mode)
 (font-lock-mode)
 (global-eldoc-mode)
 (global-font-lock-mode)
 (global-undo-tree-mode)
 (hs-minor-mode)
 (line-number-mode)
 (menu-bar-mode)
 (mouse-wheel-mode)
 (shell-dirtrack-mode)
 (show-paren-mode)
 (tool-bar-mode)
 (tooltip-mode)
 (transient-mark-mode)
 (undo-tree-mode)
 (xterm-mouse-mode))
(disabled-minor-modes
 (abbrev-mode)
 (auto-compile-mode)
 (auto-compile-on-save-mode)
 (auto-fill-function)
 (auto-save-visited-mode)
 (buffer-read-only)
 (cl-old-struct-compat-mode)
 (column-number-mode)
 (compilation-minor-mode)
 (compilation-shell-minor-mode)
 (completion-in-region-mode)
 (defining-kbd-macro)
 (diff-auto-refine-mode)
 (diff-minor-mode)
 (dired-hide-details-mode)
 (doc-view-minor-mode)
 (doc-view-presentation-mode)
 (eldoc-mode)
 (electric-layout-mode)
 (electric-quote-mode)
 (evil-org-mode)
 (flyspell-mode)
 (global-prettify-symbols-mode)
 (global-reveal-mode)
 (global-visual-line-mode)
 (gnus-dead-summary-mode)
 (gnus-undo-mode)
 (horizontal-scroll-bar-mode)
 (ido-everywhere)
 (image-minor-mode)
 (isearch-mode)
 (ispell-minor-mode)
 (jit-lock-debug-mode)
 (mail-abbrevs-mode)
 (mml-mode)
 (next-error-follow-minor-mode)
 (org-cdlatex-mode)
 (org-src-mode)
 (org-table-follow-field-mode)
 (orgtbl-mode)
 (outline-minor-mode)
 (overwrite-mode)
 (paragraph-indent-minor-mode)
 (prettify-symbols-mode)
 (racket-smart-open-bracket-mode)
 (racket-xp-mode)
 (rectangle-mark-mode)
 (reveal-mode)
 (sh-electric-here-document-mode)
 (size-indication-mode)
 (tab-bar-history-mode)
 (tab-bar-mode)
 (temp-buffer-resize-mode)
 (undo-tree-visualizer-selection-mode)
 (unify-8859-on-decoding-mode)
 (unify-8859-on-encoding-mode)
 (url-handler-mode)
 (use-hard-newlines)
 (view-mode)
 (visible-mode)
 (visual-line-mode)
 (window-divider-mode))
```
</details>

**Comment by greghendershott** [comment]
15-20 seconds is not normal. That's about 10X longer than it takes for me with say Emacs 25.2.2 and Ubuntu 18.04 "native" on an X1 Carbon 6th gen.

It's been awhile since I smoke-tested on WSL. I will try to find time to take it for a spin, soon. (If my experience is as slow as yours, I'll investigate and try to fix. Otherwise, I'll probably have some follow-up questions including what flavor of WSL you're using, and so on.)

**Comment by sailikk** [comment]
Okay, thank you. I am running Emacs 27.1.

I probably should also have mentioned that I am using Emacs as a GUI window, and am running WSL 2. WSL 2 has had a few issues with Linux GUIs, so I did test it out with terminal emacs (`emacs -nw`), but I see the exact same startup time, so perhaps the issue isn't with the GUI connection.

**Comment by greghendershott** [comment]
What I saw just now:

- `uname -r` says `4.4.0-17763-Microsoft`

- `sudo apt install emacs racket`

Installs Emacs 25.2.2 and Racket 6.11.

- `emacs -nw`

- `M-x find-file` `/tmp/foo.rkt`, which is simply

```racket
#lang racket/base
"hello"
```

- <kbd>C-c C-c</kbd>

It takes about 8-10 seconds before the REPL window appears.

- If I enter `(exit)` at the REPL, go back to `foo.rkt` buffer, and <kbd>C-c C-c</kbd> again, this time it opens "instantly".

---

As best I can tell, the initial 8-10 second delay is mostly in starting up the Racket Mode "back end", and getting to the point where its ephemeral TCP port for REPL session connections is available. Making the connection seems quite fast.

8-10 seconds:

- Seems about 2X better than you're seeing.
- I'm not shocked it's slower than I see on native/raw Ubuntu? (Windows in general is slower opening files. Although I can't keep up with Microsoft's evolving WSL designs, I know at least some of them are still subject to being slow at things Windows is slow at.)

Anyway...

---

Next I tried doing `M-x` [`racket-mode-start-faster`](https://www.racket-mode.com/#racket_002dmode_002dstart_002dfaster). This took longer to run than I expected. Which in a way was encouraging, because....

Now I did `M-x racket-stop-backend`, went back to `foo.rkt`, and did <kbd>C-c C-c</kbd>. Now it took maybe 1-2 seconds. Yay!

---

**TL;DR: Try using `M-x` [`racket-mode-start-faster`](https://www.racket-mode.com/#racket_002dmode_002dstart_002dfaster)**.

To the extent WSL is just slower at opening files (maybe?), and the expansion of Racket code involves opening many files (definitely), then prepaying that by making Racket bytecode will help.  The only catch is explained at that link, you'll need to re-run that after updating Racket Mode from MELPA.

**Comment by sailikk** [comment]
Wow! That option did pique my interest when having a cursory look over the docs but man is this much better! Thank you for the help. Although we didn't really figure out why my machine (Inspiron 15 7000 series) is doing 2X worse, I don't think it matters so much lol. I will close.

**Comment by greghendershott** [comment]
Glad to hear that!
